src/node_jsvmapi.cc

```diff
@@ -465,22 +465,58 @@ namespace v8impl {
 
 }  // end of namespace v8impl
 
-// We define a macro here because, with full error handling, we may want to
-// perform additional actions at the top of our public APIs, such as bail,
-// returning an error code at the very top of each public API in the case that
-// there's already an exception pending in v8impl::TryCatch::lastException
+   // We define a macro here because, with full error handling, we may want to
+   // perform additional actions at the top of our public APIs, such as bail,
+   // returning an error code at the very top of each public API in the case that
+   // there's already an exception pending in v8impl::TryCatch::lastException
 #define NAPI_PREAMBLE(e) \
```

Comment:
I think we should clear the error code at the top. Every API call should start with tabula rasa and should short-circuit if there's an exception pending. If everything is fine, it should be sufficient to just `return napi_ok;` without having to do an epilogue. For everything else, there's `RETURN_VALUE_IF_FAIL()`.

... and if we go with our existing signatures and not with `napi_errorvalue` return values, we can simply intersperse `RETURN_VALUE_IF_FAIL(something, nullptr|false|0)` throughout the code.